crates/ruff/src/args.rs

```diff
@@ -745,38 +745,34 @@ fn resolve_bool_arg(yes: bool, no: bool) -> Option<bool> {
     }
 }
 
+/// Enumeration of various errors that could occur
+/// when trying to parse a --config CLI flag value
 #[derive(Debug)]
-enum TomlParseFailureKind {
-    SyntaxError,
-    UnknownOption,
+enum InvalidConfigFlagError {
+    InvalidToml(toml::de::Error),
+    /// It was valid TOML, but not a valid ruff config file.
+    /// E.g. the user tried to select a rule that doesn't exist,
+    /// or tried to enable a setting that doesn't exist
+    ValidTomlButInvalidRuffSchema(toml::de::Error),
+    /// It was a valid ruff config file, but the user tried to pass a
+    /// value for `extend` as part of the config override.
+    // `extend` is special, because it affects which config files we look at
+    /// in the first place. We currently only parse --config overrides *after*
+    /// we've combined them with all the arguments from the various config files
+    /// that we found, so trying to override `extend` as part of a --config
+    /// override is forbidden.
+    ExtendPassedViaConfigFlag,
 }
 
-impl std::fmt::Display for TomlParseFailureKind {
-    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        let display = match self {
-            Self::SyntaxError => "The supplied argument is not valid TOML",
-            Self::UnknownOption => {
+impl InvalidConfigFlagError {
```

Comment:
Nit: Implementing `Display` is mandatory for proper "Error" types that also implement the `Error` trait. That's why I liked the old implementation more (might also be worth to implement `impl Error for InvalidConfligFlagError`). 

There's still a way for you to get to the error message by using `error.to_string()` (`to_string` is implemented by all types that implement `Display`. 

The downside is that it possibly requires an allocation. I think that's fine, considering that we generate at most one error ;)

Reply:
So, the reason why I got rid of the `Display` implementation is because I wanted to make invalid states unrepresentable in `ConfigArgumentParser::parse_ref`. With the new code structure, it's impossible to proceed beyond line 866 if we have an enum variant that doesn't have an underlying error -- and that's good, because those need to be treated differently when describing them in words than the enum variants that do have underlying errors. But if I implemented `Display` for this enum, it would work for all the variants alike, which actually feels like a bit of a footgun here.

Really this enum isn't an error type at all (we never wrap it with an `Err` and return it), just an enumeration of the various kinds of failure modes we could have had when trying to parse the a `--config` flag as TOML. Possibly it should just be renamed... but I'm still pretty new to Rust error-handling patterns :) What do you think?

Reply:
I see. Maybe name it `InvalidConfigFlagReason` to make it clear that it's never used in an `Err` (which I assumed is the case)